bam_sort.c

```diff
@@ -126,26 +126,30 @@ typedef struct {
     bam1_p b;
 } heap1_t;
 
-#define __pos_cmp(a, b) ((a).pos > (b).pos || ((a).pos == (b).pos && ((a).i > (b).i || ((a).i == (b).i && (a).idx > (b).idx))))
-
-static inline int bam1_lt_by_tag(const bam1_p a, const bam1_p b);
+static inline int bam1_cmp_by_tag(const bam1_p a, const bam1_p b);
 
 // Function to compare reads in the heap and determine which one is < the other
 static inline int heap_lt(const heap1_t a, const heap1_t b)
 {
     if (g_is_by_tag) {
         int t;
         if (a.b.b == NULL || b.b.b == NULL) return a.b.b == NULL? 1 : 0;
-        t = bam1_lt_by_tag(b.b,a.b);
-        return t;
+        t = bam1_cmp_by_tag(a.b, b.b);
+        if (t != 0) return t > 0;
     } else if (g_is_by_qname) {
-        int t;
+        int t, fa, fb;
         if (a.b.b == NULL || b.b.b == NULL) return a.b.b == NULL? 1 : 0;
```

Comment:
This line of code is duplicate for both cases. So we could put it above the if/else block. Also, I would rewrite it as:
```
if (a.b.b == NULL)
    return 1;
if (b.b.b == NULL)
    return 0;
```
In this way, there are at most two comparisons, while previously there were at least two.